site/content/posts/2021-06-21-windows-e2e.md

```diff
@@ -0,0 +1,134 @@
+
+To get around this, we are going to specify that the e2e plugin run on a Linux node. In our simple, 2-node cluster, this means that we need to add a nodeSelector to the podSpec as well as tolerate the control-plane NoSchedule taint (no need to actually run these commands, we are just discussing the logic right now).
+
+## Problem 2: Windows Skipped By Default
+
+Now if we were to run our chosen test with the tolerations/nodeSelectors we need, Sonobuoy will get placed on the Windows node and the e2e plugin will run from the Linux control-plane node. However, now you’ll run into another blocker: the tests are explicitly coded to skip any Windows test by default unless a flag is specified. Looking through the code, you can see that we just need to provide the `--node-os-distro=windows` flag when invoking the tests.
+
+If you include this value in the plugins environment variable `E2E_EXTRA_ARGS`, combined with the other points discussed above, you canb successfully run and pass a Windows specific test in your Kubernetes cluster.
+
+But running just one test on a simple test cluster isn't very useful. Let's talk about how we can run more tests on more complex clusters.
+
+## Milestone Two: More Tests!
+
+If you try and simply run all of the conformance tests (the default `E2E_FOCUS` when running Sonobuoy) you will still hit some tests that don’t work well on Windows either by design or due to the early development stage of the Windows tests. As a goal, let’s copy what the upstream sig-windows tests execute and run the tests matching the focus/skip regular expressions:
+
+```
+focus=\[Conformance\]|\[NodeConformance\]|\[sig-windows\]|\[sig-apps\].CronJob|\[sig-api-machinery\].ResourceQuota|\[sig-scheduling\].SchedulerPreemption|\[sig-autoscaling\].\[Feature:HPA\]  
+
+skip=\[LinuxOnly\]|\[Serial\]|GMSA|Guestbook.application.should.create.and.stop.a.working.application
+```
+
+Now you’re targeting the correct set of tests, but you’ll eventually get failures such as:
+
+```
+no matching manifest for windows/amd64 in the manifest list entries
+```
+
+This is because although we have a Sonobuoy Windows image and placed the e2e pod on a Linux node, the e2e tests themselves launch other workloads which use various images. The default images that upstream Kubernetes tests use are not all Windows compatible yet.
+
+To override this we have to tell the tests where to look up test images. This is the same process used to test air-gapped clusters. Typically, users pass this information to Sonobuoy by providing a YAML file which details where the various image repos are and setting a command-line flag. However, Sonobuoy v0.52.0 adds the ability to code ConfigMap information into the plugin definition itself. This means that we will handle coding it into the plugin with no extra work on your part. Luckily for us, [Windows-testing][windowsTesting] provides a matrix of Windows/Kubernetes versions and their respective requirements.
```

Comment:
Suggestion: `This means that we will handle coding it into the plugin with no extra work on your part...` -> `This means the plugin will handle automatically handle the list with no extra work on...`

Reply:
```suggestion
To override this, we have to tell the tests where to look up test images. This is the same process used to test air-gapped clusters. Typically, users pass this information to Sonobuoy by providing a YAML file which details where the various image repos are and setting a command-line flag. However, Sonobuoy v0.52.0 adds the ability to provide ConfigMap information into the plugin definition itself. This means the plugin will automatically handle the list with no extra work on your part. Luckily for us, [Windows-testing][windowsTesting] provides a matrix of Windows/Kubernetes versions and their respective requirements.
```

Reply:
included @vladimirvivien's suggestion too